### PR TITLE
Automate serverless deployment

### PR DESCRIPTION
**Pull request by tvararu**
This updates the `deploy.sh` script to work on Travis. When someone pushes to master, it will automatically deploy a new version of the scripts to production.

TODO:

- [x] Update .travis.yml
- [x] Update README

### CODE REVIEW

File: picker-data-_star_-eu-west-2-policy.json [hunk 1]
```diff
@@ -0,0 +1,183 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "cloudformation:List*",
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: I feel like this is probably too broad given the `*` resource. We don't do this but you can imagine a scenario where a CloudFormation stack might export some variables that you might not want to be accessible by a third-party.

https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_ListExports.html

File: picker-data-_star_-eu-west-2-policy.json [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+      "Effect": "Allow",
+      "Action": [
+        "s3:CreateBucket"
+      ],
+      "Resource": [
+        "arn:aws:s3:::*"
+      ]
+    }
+  ]
+}
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Newline at the end of the file?

File: picker-data-_star_-eu-west-2-policy.json [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        "*"
+      ]
+    },
+    {
+      "Action": [
+        "logs:CreateLogGroup",
+        "logs:CreateLogStream"
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:*:*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Aren't these log groups namespaced by the Lambda function so we can tighten up which resources it applies to?

File: picker-data-_star_-eu-west-2-policy.json [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      ],
+      "Effect": "Allow"
+    },
+    {
+      "Action": [
+        "logs:PutLogEvents"
+      ],
+      "Resource": [
+        "arn:aws:logs:eu-west-2:*:*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Same as above for this?

File: picker-data-_star_-eu-west-2-policy.json [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+      "Action": "kinesis:*",
+      "Resource": [
+        "arn:aws:kinesis:*:*:stream/picker-data-*-eu-west-2"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": "iam:*",
+      "Resource": [
+        "arn:aws:iam::*:role/picker-data-*-eu-west-2-lambdaRole"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Where does this role come from?

File: picker-data-_star_-eu-west-2-policy.json [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        "arn:aws:apigateway:*::/restapis*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "iam:PassRole"
+      ],
+      "Resource": [
+        "arn:aws:iam::*:role/*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: It doesn't seem like a good idea that we can PassRole to any Role in our account.

File: picker-data-_star_-eu-west-2-policy.json [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        "arn:aws:events:*:*:rule/picker-data-*-eu-west-2"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:CreateBucket"
+      ],
+      "Resource": [
+        "arn:aws:s3:::*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: I take it the S3 bucket name is something random?

File: picker-data-_star_-eu-west-2-policy.json [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        "arn:aws:sqs:*:*:picker-data-*-eu-west-2"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "cloudwatch:GetMetricStatistics"
+      ],
+      "Resource": [
+        "*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Do the resources here not have similarly namespaced things as with `logs:*`?

File: picker-data-_star_-eu-west-2-policy.json [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "lambda:Get*",
+        "lambda:List*",
+        "lambda:CreateFunction"
+      ],
+      "Resource": [
+        "*"
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: Can this not be `arn:aws:lambda:eu-west-2:*:function:picker-data-*-*` for `lambda:Get*`?

**Comment by tvararu** [comment]
@samcrang can you give this another look? I've significantly slimmed down all the permissions.

I've scoped everything down apart from `cloudformation:ValidateTemplate` which seems to ignore ARN scoping.

`apigateway:*` are the only ones I couldn't really scope further as their ARNs look like `arn:aws:apigateway:eu-west-2::/restapis/f8rw5ex6ma/deployments`.

(I will rebase and clean the commits if all checks out)

File: picker-data-_star_-eu-west-2-policy.json [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        "lambda:CreateFunction"
+      ],
+      "Resource": [
+        "arn:aws:lambda:eu-west-2:*:function:picker-data-*-*"
+      ]
+    },
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:CreateBucket",
```
Comment (samcrang) [inline picker-data-_star_-eu-west-2-policy.json]: I doubt you need `CreateBucket`.

Comment (tvararu) [reply picker-data-_star_-eu-west-2-policy.json]: I think you're right!

**Comment by tvararu** [comment]
@samcrang updated again!

**Review by samcrang — APPROVED** [review]
Looks reasonable m8. Assuming you rebase the whole thing.

**Comment by tvararu** [comment]
Thanks @samcrang 👍

**Pull request merged**